### PR TITLE
Add favoriting

### LINKED ISSUE #48: Favorite Image
#### As a logged in user I want to favorite an image.

Acceptance criteria:
- [ ] Each image on the index page has a "favorite" button.
- [ ] The button indicates how many users have favorited that image.
- [ ] When you favorite an image the button state changes to a "favorited"
  state.
- [ ] When you favorite an image the count of users who have favorited that
  image is updated.
- [ ] If you have already favorited an image, clicking the button "unfavorites"
  the image.
- [ ] Favoriting/unfavoriting is done via AJAX.

Discussion topics:
- [ ] What can we do to cache the favorites count?
- [ ] How can we make our images relation for the index page include whether
  the image was favorited by the current user?
- [ ] How can we develop the front end and back end separately?
- [ ] How can we reuse templates on the server and client?
- [ ] What are the different types of errors we should handle?

Reference material:
- http://guides.rubyonrails.org/association_basics.html#counter-cache
- http://guides.rubyonrails.org/active_record_querying.html#scopes
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template

Dependencies:
- Authorization

### PR DESCRIPTION
**Pull request by rmacklin**
Closes #48

### CODE REVIEW

File: db/migrate/20160226050358_create_favorites.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class CreateFavorites < ActiveRecord::Migration
+  def change
+    create_table :favorites do |t|
+      t.integer :user_id, null: false
+      t.integer :image_id, null: false
+
+      t.timestamps null: false
+    end
+
+    add_column :images, :favorites_count, :integer, default: 0
```
Comment (bboe) [inline db/migrate/20160226050358_create_favorites.rb]: Should this also be `null: false`?

File: db/schema.rb [hunk 2]
```diff
@@ -11,13 +11,24 @@
...
+  end
+
+  add_index "favorites", ["image_id"], name: "index_favorites_on_image_id"
+  add_index "favorites", ["user_id"], name: "index_favorites_on_user_id"
+
+  create_table "images", force: :cascade do |t|
+    t.string   "url",                         null: false
+    t.datetime "created_at",                  null: false
+    t.datetime "updated_at",                  null: false
     t.integer  "user_id"
```
Comment (bboe) [inline db/schema.rb]: This comment probably belongs on the authorization story, but should we enforce that all images have a user?

File: db/migrate/20160226050358_create_favorites.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+    add_index :favorites, :image_id
```
Comment (bboe) [inline db/migrate/20160226050358_create_favorites.rb]: Do we need these to be separate indexes? It seems the two common actions are:
* has my user favorited a certain image (compound index on both fields)
* how many users have favorited a certain image (index on image_id)

It seems like those two queries can be encompassed by a single index across both fields with image_id first. One use-case I can think of is having a single-index on `user_id` would be give me all images I've favorited.

Now that I think about it, maybe that's used when checking the list of front-page image to see which have been favorited by the current user.

File: db/migrate/20160226050358_create_favorites.rb (hunk 1 above)
Comment (rmacklin) [inline db/migrate/20160226050358_create_favorites.rb]: Fixed in b441950bdc0e2e274747bce2b4d7e14ad2131f84

File: db/schema.rb (hunk 2 above)
Comment (rmacklin) [inline db/schema.rb]: Hmmm, I could've sworn I had made this one `null: false` in that story. I think it must have been before I redid the migration using [`add_reference`](https://github.com/AppFolioOnboarding/vanilla/commit/27baedb2928a536304875d5c515934cfdebb2fd5#diff-60f9cda433d6e0cb14f7e416c1d350e6R3). In any case, fixed in b441950bdc0e2e274747bce2b4d7e14ad2131f84.

File: db/migrate/20160226050358_create_favorites.rb (hunk 3 above)
Comment (rmacklin) [inline db/migrate/20160226050358_create_favorites.rb]: Hmm, good question. I had originally planned on implementing the "give me all images I've favorited" action you mentioned, but cut it out to trim the story scope. But I left the index on `user_id` for [this line](https://github.com/AppFolioOnboarding/vanilla/blob/815b2f52777fb9bd44c8e085e5043648e5735ddf/app/models/image.rb#L14), though I think you're right that a compound index would've worked there.

**Pull request merged**